src/shared/sam/cli/defaultValidatingSamCliProcessInvoker.ts

```diff
@@ -0,0 +1,75 @@
+/*!
+ * Copyright 2018 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+'use strict'
+
+import {
+    DefaultSamCliProcessInvoker,
+    resolveSamCliProcessInvokerContext,
+    SamCliProcessInvokerContext
+} from './samCliInvoker'
+import { notifySamCliValidation } from './samCliValidationNotification'
+import {
+    DefaultSamCliValidator,
+    InvalidSamCliError,
+    InvalidSamCliVersionError,
+    SamCliNotFoundError, SamCliValidator,
+    SamCliVersionValidation
+} from './samCliValidator'
+
+/**
+ * Validates the SAM CLI version before making calls to the SAM CLI.
+ */
+export class DefaultValidatingSamCliProcessInvoker extends DefaultSamCliProcessInvoker {
```

Comment:
I'm not a fan of this inheritance structure. `DefaultValidatingSamCliProcessInvoker` adds validation to `DefaultSamCliProcessInvoker`, but `DefaultSamCliProcessInvoker` needs to be aware of validation/have validation hooks? Additionally, it's easier to read the code when it's not spread across two files/classes. What do you think about something like:

```typescript
export class DefaultValidatingSamCliProcessInvoker implements SamCliProcessInvoker {
    public constructor(
        context: SamCliProcessInvokerContext = resolveSamCliProcessInvokerContext(),
        private readonly validator: SamCliValidator = ...,
        private readonly invoker: SamCliProcessInvoker = ...,
    ) {
    }
    ...
}
```

Reply:
I was thinking about this after the fact as well. Let me explore this a little.